src/lib/coerce.js

```diff
@@ -19,21 +19,70 @@ var nestedProperty = require('./nested_property');
 var counterRegex = require('./regex').counter;
 var modHalf = require('./mod').modHalf;
 var isArrayOrTypedArray = require('./array').isArrayOrTypedArray;
+var isTypedArraySpec = require('./array').isTypedArraySpec;
+var decodeTypedArraySpec = require('./array').decodeTypedArraySpec;
+var coerceTypedArraySpec = require('./array').coerceTypedArraySpec;
+
 
 exports.valObjectMeta = {
     data_array: {
         // You can use *dflt=[] to force said array to exist though.
         description: [
             'An {array} of data.',
-            'The value MUST be an {array}, or we ignore it.',
-            'Note that typed arrays (e.g. Float32Array) are supported.'
+            'The value could be an {array}',
+            'noting that typed arrays (e.g. Float32Array) are also supported.',
+            'It could also be an object in the form of',
+            'v: {, dtype: \'float32\', bvals: [/* ... */]}, shape: [dim0 (, dim1, (dim3))]',
```

Comment:
`dtype`, `bvals` are strings.
I think there are benefits for shape to be a string (or alternatively a single number) but not an array.
Then we could `split` shape string and convert it to an array attached to `coerced` cache.
Some examples for shape:
```js
shape: 1000 // number
shape: '1000' // numeric string
shape: '10 | 100' // 2d array
shape: '10 | 10 | 10' // 3d array
```

Reply:
And maybe instead of an object we could simply pass one string like this:
```js
"data:array/2/3;float32,AAAAAAAAgD8AAABAAABAQAAAgEAAAKBA"
```

Reply:
I guess I don't see a benefit to using a string for the shape in the public interface, or encoding everything into a single string. 

For producers, they will generate the base64 encoding as an individual string from the array memory. This is often going to be really large so I wouldn't want to force them to concatenate it to produce another very large string.

For shape, again producers will always have the shape in a list/array form, so converting to a special string encoding would be an extra step.  I understand the idea of using the shape as a string as part of the cache, and we could generate that if we need to as part of the cache store/lookup process.  But, the design right now is to only cache the `base64` to `ArrayBuffer` values.  I thought this would be more flexible that storing the full arrays/typed arrays in the cache, and it removes the need to use shape in the cache.

Reply:
To avoid string cocats, one could add a second argument in decode function of the module here
https://github.com/niklasvh/base64-arraybuffer/blob/29d0f8204348d913e5136f0c713a510c5a1da574/lib/base64-arraybuffer.js#L39 
so that it start the process after nth character of the string allocated for the header.

Reply:
FYI - just opened this PR: https://github.com/niklasvh/base64-arraybuffer/pull/25.

Reply:
And you could simply pick https://github.com/plotly/plotly.js/commit/67c60e3e5e60f1f9ac05d54d8023e154cef1c0a8, 
if you want to play with it.

Reply:
One possible way forward would be to have e.g. `x` be a base64 string, and `xsrc` contain a stringified version of the 'header' information. That way we don't have to accept objects anywhere and create new keys etc.

Reply:
This sounds better, but I guess I'm still not clear on what problem we're trying to solve here.

Reply:
Resolved.